src/test/socket/socket/test_socket.rs

```diff
@@ -0,0 +1,501 @@
+        //libc::AF_TIPC,
+        //libc::AF_BLUETOOTH,
+        //libc::AF_ALG,
+        //libc::AF_VSOCK,
+        //libc::AF_XDP,
+        0xABBA,
+    ];
+    // since the behaviour when using SOCK_RAW or SOCK_PACKET depends on the user's privileges,
+    // we don't test them
+    let sock_types = [
+        libc::SOCK_STREAM,
+        libc::SOCK_DGRAM,
+        libc::SOCK_SEQPACKET,
+        //libc::SOCK_RAW,
+        libc::SOCK_RDM,
+        //libc::SOCK_PACKET,
+        0xABBA,
+    ];
+    let flags = [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC, 0xABBA];
+    let protocols = [0, libc::IPPROTO_TCP, libc::IPPROTO_UDP];
+
+    // product of the sets of arguments
+    let mut tests = Vec::new();
+
+    for socket_fn in [SocketFn::Socket, SocketFn::Syscall].iter() {
+        for domain in domains.iter() {
+            for sock_type in sock_types.iter() {
+                for flag in flags.iter() {
+                    for protocol in protocols.iter() {
+                        let args = SocketArguments {
+                            domain: *domain,
+                            sock_type: *sock_type,
+                            flag: *flag,
+                            protocol: *protocol,
+                        };
+                        tests.push((*socket_fn, args));
+                    }
+                }
+            }
+        }
+    }
+
+    // add any shadow-passing tests we are missing
+    for passing_test in get_passing_tests().iter() {
+        if !tests.contains(passing_test) {
+            tests.push(*passing_test);
+        }
+    }
+
+    tests
+}
+
+fn run_tests<'a, I>(
+    tests: I,
+    error_conditions: &[ErrorCondition],
+    summarize: bool,
+) -> Result<(), ()>
+where
+    I: Iterator<Item = &'a (SocketFn, SocketArguments)>,
+{
+    // the set of file descriptors returned
+    let mut used_fds = std::collections::HashSet::new();
+
+    // assume that we haven't closed 0, 1, or 2
+    used_fds.insert(0);
+    used_fds.insert(1);
+    used_fds.insert(2);
+
+    // number of tests that passed, and the total performed
+    let mut num_passed = 0;
+    let mut num_performed = 0;
+
+    // run the tests
+    for (socket_fn, args) in tests {
+        // by default, we don't expect to fail or want to check the errno
+        let mut should_error = false;
+        let mut expected_errno = None;
+
+        // check the failure conditions
+        for cond in error_conditions.iter() {
+            // if the arguments match one of the failure conditions
+            if args.matches(cond) {
+                should_error = true;
+                expected_errno = cond.expected_errno;
+                break;
+            }
+        }
+
+        print!("Testing {}...", socket_fn.format(args));
+
+        num_performed += 1;
+
+        let fd;
+        match check_socket_call(args, *socket_fn, should_error, expected_errno) {
+            // the test passed
+            Ok(result_fd) => {
+                println!(" ✓");
```

Comment:
Neat check mark :)

Reply:
I'm sure this will break someone's editor at some point :)